### PR TITLE
[FLINK-10747] Update SingleInputGate#requestedPartitionsFlag under lock

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

In order to avoid a race condition between SingleInputGate#requestPartitions and
SingleInputGate#updateInputChannel we need to change the field requestedPartitionsFlag
under the requestLock.

This PR fixes a problem introduced by #6974.

## Verifying this change

- Manually tested

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by zhijiangW** [comment]
If we want to avoid the overhead by calling `requestPartitions` on each `getNextBufferOrEvent`, we can change to trigger the request in the construction of `InputGate`. Then it will be called only once for initialization to avoid additional check each buffer level. Maybe it is not the scope of this issue.  :)

**Comment by StefanRRichter** [comment]
Fix looks good to me 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 1]
```diff
@@ -482,23 +482,25 @@ public void requestPartitions() throws IOException, InterruptedException {
 		}
 
 		synchronized (requestLock) {
-			if (isReleased) {
-				throw new IllegalStateException("Already released.");
-			}
+			if (!requestedPartitionsFlag) {
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: we should follow the same "having the shorter/simpler branch first" approach as before (also see https://github.com/apache/flink/pull/6974#discussion_r229666072 )

**Comment by tillrohrmann** [comment]
I talked to @pnowojski and @NicoK offline and we concluded that the best option would be to revert the changes and to revisit the problem of the `requestLock` from a bit different angle [FLINK-10762](https://issues.apache.org/jira/browse/FLINK-10762).

**Pull request closed without merging**